fix: support cases when user-provided queries have 'returning'

vlsi opened the pull request:
This change includes: drop v2 protocol support, and query parsing refactoring.

Currently query parse cache is still per-connection, however "returningColumNames" are part of cache key, thus the parse cache can be made global.

This fixes #488

davecramer commented on pgjdbc/src/main/java/org/postgresql/core/CachedQueryCreateAction.java:
```diff
@@ -0,0 +1,68 @@
+/*-------------------------------------------------------------------------
+*
+* Copyright (c) 2015, PostgreSQL Global Development Group
```
any reason not to use 2016 here ?

vlsi replied:
Will add a checkstyle rule to validate java headers

davecramer commented on pgjdbc/src/main/java/org/postgresql/PGProperty.java:
```diff
@@ -357,6 +357,14 @@
+   *
+   * This mode is meant for debugging purposes and/or for cases when extended protocol cannot be used (e.g. logical replication protocol)
+   */
+  SIMPLE_PROTOCOL_ONLY("simpleProtocolOnly", "false",
```
we need to add this to the docs

davecramer commented on pgjdbc/src/main/java/org/postgresql/core/BaseQueryKey.java:
```diff
@@ -0,0 +1,47 @@
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
```
is there an optimizer reason why not to check this first ? Or do you expect the first == clause to be more frequent ?

vlsi replied:
This is automatically generated by IntelliJ IDEA, so I assume it to be just fine.

davecramer replied:
Fair enough, I was just curious

Gordiychuk commented:
I review part that add ability execute query via simple query protocol and I have one suggestion:

Let's change property name for enable simple query protocol to property that can accept not only true or false. A first mode will be execute simple query protocol  everywhere. A second of property mode will use simple query protocol only for Statement(not PreparedStatement).  A third mode will same previous behavior use extended query protocol everywhere and will be chose by default. 

By introduce this modes user can more conveniently configure connection and for example where overhead of extended query protocol not applicable user will use Statement, but where necessary use features like bind variables, caching etc user will use PreparedStatement. With the current changes user should use 2 connection to database for reproduce same behavior.

Gordiychuk commented:
Maybe split `org.postgresql.core.v3.QueryExecutorImpl` to few classes to avoid many checks like this
```java
if ((flags & QueryExecutor.QUERY_EXECUTE_AS_SIMPLE) != 0) {
          // Sync message is not required for 'Q' execution as 'Q' ends with ReadyForQuery message
          // on its own
        } else {
          sendSync();
        }
```

vlsi commented:
>A first mode will be execute simple query protocol everywhere

This is what `simpleProtocolOnly=true` means

>A second of property mode will use simple query protocol only for Statement(not PreparedStatement)

This is what `simpleProtocolOnly=false` (the default) means.
Note that I've added QUERY_EXECUTE_AS_SIMPLE to PgStatement.execute(sql) and non-prepared batch execute.

The good part with this mode is pgjdbc can do much less parsing for `createStatement(String sql)`. For instance, it does not need to replace `?` and split statements by `;` as simple 'Q' execute supports multi-statements natively at the database.

>A third mode will same previous behavior use extended query protocol everywhere and will be chose by default.

That's a good question.
Technically speaking, `conneciton.createStatement()` cannot use binds, so the only benefit of using extended query I can imagine is:
1) Avoid sending query text (when the same text for createStatement is reused).
2) Use binary format for receive.

Both of the above happen only in case server-prepared statement is used. And that requires "5 executions of the same query text for the server-prepared statement to be activated".
I've no idea how often that happens.

vlsi commented:
I've changed `simpleProtoclOnly=true|false` into `preferQueryMode=extended (default) | simple | extendedForPrepared`.
Better naming suggestions are welcome if any.

* extended (default): send all queries as extended (parse, bind, exec)
* simple: send all queries as simple 'Q'
* extendedForPrepared: send prepared statements as extended, and non-prepared statements as 'Q'

vlsi commented:
Just some more comments for "simple vs extended".
I think it is unlikely for "simple batch" to be used with "repeated statements".
Typical usage I can think of is "insert into t(x) values(1)", "insert into t(x) values(2)", etc, so query caching would not help there.

I think "force simple query" mode should be used for `java.sql.Statement#addBatch(String sql)`.
